fix: This fixes the issue with recent 2023 update where FindObjectsOfType was renamed to FindObjectsByType

NoelStephensUnity opened the pull request:
This fixes the issue with recent 2023 update where FindObjectsOfType was renamed to FindObjectsByType

NoelStephensUnity committed:
update

Adding the 2023.1.a21 FindObjectsByType fix for testproject.

NoelStephensUnity committed:
update

Adding the 2023.1.a21 FindObjectsByType fix for testproject.multiprocesstests.

NoelStephensUnity committed:
Merge branch 'develop' into fix/findobjectsoftype-to-findobjectsbytype

NoelStephensUnity committed:
update

Fixes the rest of the assemblies that had scripts that used FindObjectsOfType

NoelStephensUnity committed:
test fix

Possible fix for what seems to be a timing related issue where not destroying the source prefab NetworkObject/GameObject it could cause an exception to occur.

NoelStephensUnity committed:
Revert "test fix"

This reverts commit 420768a5552ea726f1f032d85d7c4473d0a755b6.

ashwinimurt commented on com.unity.netcode.gameobjects/TestHelpers/Runtime/com.unity.netcode.testhelpers.runtime.asmdef:
```diff
@@ -6,12 +6,20 @@
     ],
+    "autoReferenced": false,
     "defineConstraints": [
+        "UNITY_INCLUDE_TESTS",
```
Duplicate, exists below.

NoelStephensUnity replied:
Yeah... that happens when you open the dang thing in the inspector view... will fix that.

ashwinimurt commented on com.unity.netcode.gameobjects/Tests/Runtime/com.unity.netcode.runtimetests.asmdef:
```diff
@@ -11,12 +11,20 @@
     ],
+    "autoReferenced": false,
     "defineConstraints": [
+        "UNITY_INCLUDE_TESTS",
```
Duplicate, exists below

NoelStephensUnity committed:
update

Removing the duplicate UNITY_INCLUDE_TESTS

NoelStephensUnity committed:
fix

Prevent final NetworkObject clean up from causing exceptions on 2023 builds.

NoelStephensUnity committed:
test fix

This takes some additional scenarios into consideration.

0xFA11 commented on com.unity.netcode.gameobjects/Editor/NetworkManagerHelper.cs:
```diff
@@ -65,7 +65,11 @@ private static void ScenesInBuildActiveSceneCheck()
             var scenesList = EditorBuildSettings.scenes.ToList();
             var activeScene = SceneManager.GetActiveScene();
             var isSceneInBuildSettings = scenesList.Count((c) => c.path == activeScene.path) == 1;
+#if USE_FINDOBJECTSBYTYPE
```
use `#if UNITY_2023_1_OR_NEWER` macro (or a similar one) instead of our very own asmdef ones?

NoelStephensUnity replied:
It is specific to 2023.1.0a21 or newer and I couldn't find a define that already existed for just that version and up. 
Will look around again to see if there is already a define specific to that update.

0xFA11 replied:
usually, `UNITY_X_Y_OR_NEWER` macros show themselves in release (and beta) versions. if 2023 specific ones are not there, maybe this PR is "too early".

NoelStephensUnity replied:
Just pushed the pragma approach when UNITY_2023_1_OR_NEWER is defined.

NoelStephensUnity committed:
fix

This fixes the integration test issue on 2023.1.0a21.

NoelStephensUnity committed:
update

getting that section of the code back to its original

NoelStephensUnity committed:
update

Removing the extra destroyimmediate...

NoelStephensUnity committed:
style

White space

NoelStephensUnity committed:
refactor

Approaching the fix by using Jesse's suggested pragma 612 & 618 when UNITY_2023_1_OR_NEWER is defined.

NoelStephensUnity committed:
update

Reverting the asmdef files back to their original state prior to this PR.

NoelStephensUnity committed:
style

adding new line

NoelStephensUnity committed:
style

remove one new line add another

NoelStephensUnity committed:
update - clean up

Cleaning up the adjustments so that as little code and how it is formatted is impacted (as possible).

0xFA11 commented on com.unity.netcode.gameobjects/Editor/NetworkManagerHelper.cs:
```diff
@@ -65,7 +65,13 @@ private static void ScenesInBuildActiveSceneCheck()
             var scenesList = EditorBuildSettings.scenes.ToList();
             var activeScene = SceneManager.GetActiveScene();
             var isSceneInBuildSettings = scenesList.Count((c) => c.path == activeScene.path) == 1;
+#if UNITY_2023_1_OR_NEWER
```
I think we should make it work with the new API after 2023_1 and use the now obsolete API if below 2023_1 and pragma disable warnings/errors. I will reach out to you on Slack later to discuss.

NoelStephensUnity committed:
style

removing extra lines that were not there prior to this PR.

NoelStephensUnity committed:
update

The new direction is to just simply use UNITY_2023_1_OR_NEWER.

NoelStephensUnity committed:
update

Modifying a few of the FindObjectsSortMode for areas that looked like they might require their results ordered by InstanceID.

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -2018,7 +2018,11 @@ internal void PopulateScenePlacedObjects(Scene sceneToFilterBy, bool clearSceneP
             }
 
+#if UNITY_2023_1_OR_NEWER
+            var networkObjects = UnityEngine.Object.FindObjectsByType<NetworkObject>(FindObjectsSortMode.InstanceID);
```
<3

0xFA11 replied:
I like how there's sort by InstanceId on the API built-in.
